### PR TITLE
VERIFY-XDP-VF-HOT-REMOVE: Verify XDP with VF hot removal and readd

### PR DESCRIPTION
**Pull request by abhimarathe**
While running XDP check if VF hot removal and re-add works correctly and does not impact XDP. 
While VF is removed the Synthetic driver should handle XDP program execution.

```
[LISAv2 Test Results Summary]
Test Run On           : 10/02/2020 18:37:22
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:7

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-VF-HOT-REMOVE                                                          PASS                 5.94
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, TestLocation: eastus2, Kernel Version: 5.4.0-1026-azure
```

### CODE REVIEW

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+.Description
+    This script deploys the VM, build and Install xdpdump application
+    Verify VF Hot removal and add works as expected
+#>
+
+param([object] $AllVmData,
+    [object] $CurrentTestData)
+
+# This function will add or remove VF from VM
+function Set-VF {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Please check out other bash function name convention inside. I think we used 'underscored line' or 'no symbol'

**Review by juhlee-microsoft — COMMENTED** [review]
I only have one question about bash function name. BTW, what is the target distro versions? I only see one distro test result. Is it only one version?

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 (hunk 1 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: fixed.

**Comment by abhimarathe** [comment]
To my knowledge only Ubuntu picked up XDP patches and released in kernel 5.4 which 20.04 is using. I did not here about any other distro version kernel having these patches.

**Review by juhlee-microsoft — DISMISSED** [review]

**Comment by juhlee-microsoft** [comment]
> To my knowledge only Ubuntu picked up XDP patches and released in kernel 5.4 which 20.04 is using. I did not here about any other distro version kernel having these patches.

Is there a public document for Azure XDP? I can not find any in my search.

**Comment by abhimarathe** [comment]
>Is there a public document for Azure XDP? I can not find any in my search.

No. We don't have one yet.

File: Testscripts/Linux/utils.sh [hunk 2]
```diff
@@ -3819,6 +3819,23 @@ function is_hpc_vm() {
 	fi
 }
 
+# Get name of additional synthetic interface
+function get_extra_synth_nic {
+    local ignore_if=$(ip route | grep default | awk '{print $5}')
+    local interfaces=$(ls /sys/class/net | grep -v lo | grep -v ${ignore_if})
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/utils.sh:3825]: Single grep with OR can be used.

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+.Description
+    This script deploys the VM, build and Install xdpdump application
+    Verify VF Hot removal and add works as expected
+#>
+
+param([object] $AllVmData,
+    [object] $CurrentTestData)
+
+# This function will add or remove VF from VM
+function Set_VF {
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Give appropriate name like ```Enable_Disable_VF```

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            Write-LogInfo "Starting command: $ping_command on $($receiverVMData.RoleName)"
+            Run-LinuxCmd -ip $receiverVMData.PublicIP -port $receiverVMData.SSHPort -username $user -password $password `
+                -command $ping_command -RunInBackground -runAsSudo
+            Write-LogDbg "XDP program cannot run with LRO (RSC) enabled, disable LRO before running XDP"
+            $xdp_command = "ethtool -K $iFaceName lro off && cd ~/bpf-samples/xdpdump && ./xdpdump -i $iFaceName > ~/xdpdumpout_VFTest.txt 2>&1"
+            $testJob = Run-LinuxCmd -ip $receiverVMData.PublicIP -port $receiverVMData.SSHPort -username $user -password $password `
+                -command $xdp_command -RunInBackground -runAsSudo -ignoreLinuxExitCode
+
+            Wait-Time -seconds 10
+            Set_VF $receiverVMData $false
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Capture the vf & synthetic rx counter before removing VF IFC

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            Wait-Time -seconds 20
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Increase the sleep to long values like 120 secs
Capture the rx counters for vf & synthetic ifc to check the packet reception

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            Set_VF $receiverVMData $true
+            Wait-Time -seconds 10
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1:153]: Increase the sleep to long values like 120 secs
Capture the rx counters for vf & synthetic ifc to check the packet reception

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            # Kill and verify XDP unloaded successfully
```
Comment (johnsongeorge-w) [inline Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1:165]: Conclude the test results based on the counters of rx captured at different stages.

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 10/06/2020 16:45:31
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:9

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-VF-HOT-REMOVE                                                          PASS                 8.28
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, TestLocation: eastus2, Kernel Version: 5.4.0-1026-azure


Logs can be found at C:\LISAv2\CN49\TestResults\2020-10-06-09-44-24-8407
```
From log counters:
With VF attached packet count: Synth: 815694 VF: 1012236
After VF removed packet count: Synth: 8498636
With VF attached packet count: Synth: 9865632 VF: 550842

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 (hunk 3 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Fixed

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 (hunk 4 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Fixed

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 (hunk 5 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1]: Made changes to 60 seconds.

File: Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1 (hunk 6 above)
Comment (abhimarathe) [reply Testscripts/Windows/VERIFY-VF-HOT-REMOVE-XDP.ps1:153]: fixed

File: Testscripts/Linux/utils.sh (hunk 2 above)
Comment (abhimarathe) [reply Testscripts/Linux/utils.sh:3825]: will it impact performance wise? if we keep it separated?
Just curious

**Review by johnsongeorge-w — APPROVED** [review]

**Comment by abhimarathe** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 10/07/2020 15:53:48
ARM Image Under Test  : Canonical : 0001-com-ubuntu-server-focal : 20_04-lts : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:13

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 XDP                  VERIFY-XDP-VF-HOT-REMOVE                                                          PASS                 9.74
      ARMImageName: Canonical 0001-com-ubuntu-server-focal 20_04-lts latest, Networking: SRIOV, OverrideVMSize: Standard_DS4_v2, TestLocation: eastus2, Kernel Version: 5.4.0-1026-azure


Logs can be found at C:\LISAv2\HA55\TestResults\2020-10-07-08-53-10-1525
```

**Review by juhlee-microsoft — APPROVED** [review]

**Pull request merged**